DDF-1864 Added Mock STANAG 4559 Simulator

bdeining opened the pull request:
and Initial STANAG 4559 Source and Common Framework

@kcwire 
@jrnorth 
@jlcsmith 
@dcruver 
@shane-voisard is hero
@troymohl

review-ninja commented:
<a href="https://app.review.ninja/codice/ddf/pull/541" target="_blank"><img src="https://app.review.ninja/15173563/pull/541/badge" alt="ReviewNinja"/>

jlcsmith commented on catalog/spatial/stanag4559/spatial-stanag4559-common/pom.xml:
```diff
@@ -0,0 +1,136 @@
+                    </instructions>
+                </configuration>
+            </plugin>
+            <!-- This plugin generates Java Code from IDL.  Uncomment to regenerate the STANAG 4559 IDL
```
Is there a reason not to do this with the normal build?

bdeining replied:
It doesn't seem necessary to run this plugin every time the build is run, as the IDL (and the resulting Java code) isn't likely to change.

jlcsmith commented:
It is probably worth noting that the IDL files are straight from the spec and don't need to be reviewed.

bdeining commented:
Additionally, the auto generated Java code from the idlj-maven-plugin needn't be reviewed.  (CB, GIAS, PS, UCO, UID packages).

jlcsmith commented on catalog/spatial/stanag4559/spatial-stanag4559-common/src/main/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/common/CB/CallbackHolder.java:
```diff
@@ -0,0 +1,38 @@
+/**
+* org/codice/ddf/spatial/ogc/stanag4559/catalog/common/CB/CallbackHolder.java .
+* Generated by the IDL-to-Java compiler (portable), version "3.2"
+* from /Users/bdeininger/ddf/catalog/spatial/stanag4559/spatial-stanag4559-common/src/main/idl/cb.idl
```
Any way to generate the stubs without comments like these?

bdeining commented:
Removed the generated code from the baseline and moved it into target.  @jlcsmith there doesn't seem to be an option for that plugin to omit the auto generated comments.

jrnorth commented:
:+1:

shane-voisard commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/impl/LibraryImpl.java:
```diff
@@ -0,0 +1,142 @@
+            try {
+                poa_.activate_object_with_id(manager_type.getBytes(), catalogMgr);
+            } catch (Exception e) {
+                // Ignore
```
Should not catch Exception and do noop.

stustison replied:
So two things, do not catch "Exception" and do not just ignore it. Log something out at an appropriate level. Also, I'd like to see some debug logging for large methods like this in all of these classes.

shane-voisard commented:
Successful hero build of a96f0080870b8767edce9c08bc4f205c5473f285.  Ran the mock server and mock client successfully.

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/client/Client.java:
```diff
@@ -0,0 +1,429 @@
+import org.omg.CORBA.ORB;
+
+public class Client {
+
```
I tend to instance a class from the main and then make the class variables, etc non-static.

stustison replied:
Yes. That will also make it easier to unit test this.

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/client/Client.java:
```diff
@@ -0,0 +1,429 @@
+                    BufferedInputStream inputStream =
+                            new BufferedInputStream(fileDownload.openStream());
+                    FileOutputStream outputStream = new FileOutputStream(new File(productPath));
+
```
Streams should be closed. I would use the try with resources or instancing the stream outside of the try and then checking if !=null and closing in a finally block.

stustison replied:
Definitely use try with resources where you can and then finally blocks if something doesn't implement auto closeable.

stustison commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/impl/managers/ProductMgrImpl.java:
```diff
@@ -0,0 +1,164 @@
+    public int query_availability_delay(Product prod,
+            AvailabilityRequirement availability_requirement, String use_mode)
+            throws ProcessingFault, InvalidInputParameter, SystemFault {
+        return 10;
```
What is 10?

stustison commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/impl/managers/ProductMgrImpl.java:
```diff
@@ -0,0 +1,164 @@
+
+    @Override
+    public short get_number_of_priorities() throws ProcessingFault, SystemFault {
+        return 10;
```
What is 10?

stustison commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/impl/managers/ProductMgrImpl.java:
```diff
@@ -0,0 +1,164 @@
+
+    @Override
+    public int get_default_timeout() throws ProcessingFault, SystemFault {
+        return 1;
```
1 what?

stustison commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/impl/requests/SubmitQueryRequestImpl.java:
```diff
@@ -0,0 +1,108 @@
+
+public class SubmitQueryRequestImpl extends SubmitQueryRequestPOA {
+
+    private int hits = 20;
```
Why 20?

stustison commented on catalog/spatial/stanag4559/spatial-stanag4559-common/pom.xml:
```diff
@@ -0,0 +1,135 @@
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Import-Package/>
+                        <Export-Package/>
+                        <Embed-Dependency/>
```
You don't need to list Embed-Dependency. Nothing is embedded by default.

stustison commented on catalog/spatial/stanag4559/spatial-stanag4559-common/pom.xml:
```diff
@@ -0,0 +1,135 @@
+    <properties>
+        <!-- Disable Checkstyle since the generated code won't adhere to the DDF Checkstyle rules -->
+        <checkstyle.skip>true</checkstyle.skip>
+        <findbugs.skip>true</findbugs.skip>
```
Is there some way to generate the code on the fly? A maven plugin for instance so that we don't need to check it in?

bdeining replied:
The code is no longer being checked in, so this can be removed.

stustison replied:
OK.

stustison commented on catalog/spatial/stanag4559/spatial-stanag4559-source/pom.xml:
```diff
@@ -0,0 +1,215 @@
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Import-Package/>
```
This confuses me that you have no imports, but you have dependencies and you aren't embedding anything. So how are you satisfying the dependencies?

stustison replied:
Actually from looking at the rest of the code in this. It looks like this is supposed to be standalone thing that you run from the main method? If that is the case, you shouldn't be using the maven bundle plugin at all and you should be packaging this as a jar.

bdeining replied:
The standalone code will be moved to SDK, but spatial-stanag4559-source will ultimately be a bundle like other sources.

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/client/Client.java:
```diff
@@ -0,0 +1,429 @@
+        // For each packageElement in the order response, get the parameters and
+        // related files for the product.
+        for (int i = 0; i < packageElements.length; i++) {
+            Product product = packageElements[i].prod;
```
Probably minor, but I tend to use the for each syntax vs this (e.g for PackageElement pkgElement:packageElements)

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/client/Client.java:
```diff
@@ -0,0 +1,429 @@
+
+        ProductDetails productDetails[] = {new ProductDetails(mTypes, benums, aSpec, product, "")};
+        DeliveryDetails deliveryDetails[] = {new DeliveryDetails(destination, "", "")};
+
```
Another minor style. What is the preferred place for [] on arrays. This class seems to have both:
String[] array
String array[]

stustison replied:
I believe we do String[] and I want to say that checkstyle should enforce that.

bdeining replied:
We can't use checkstyle on this package because it is mostly implementations of Java code generated from IDL, and it does not adhere to our checkstyle rules.

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/data/DAGGenerator.java:
I usually try to make my static Strings static final in case they need to be called from other places in the code.

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/data/DAGGenerator.java:
Not sure on the project preference where we want variables to look like the spec or to follow standard java coding standards. Normally CAPS with underscore are reserved for statically defined variables.

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/data/DAGGenerator.java:
These might be another place we would want to statically define these.

troymohl commented on catalog/spatial/stanag4559/spatial-stanag4559-source/src/test/java/org/codice/ddf/spatial/ogc/stanag4559/catalog/source/server/MockSTANAG4559.java:
```diff
@@ -0,0 +1,77 @@
+
+        System.out.println("Server ready...");
+        orb.run();
+    }
```
Per the ORB docs I think we need to call destroy before the process exits. Maybe add a shutdownHook to clean this up? https://docs.oracle.com/javase/7/docs/api/java/lang/Runtime.html#addShutdownHook(java.lang.Thread)

bdeining commented:
@shane-voisard 
Exceptions are now more specific and logged by the server

@troymohl 
The Client has been separated from main and instantiated there.
Streams were moved to try-with-resources.
Used For each when appropriate.
Fixed the array declaration styling.
Made most strings in the DAGGenerator static final, these will likely be used by the server later as it is built up.
Refactored some of the variables to camel case.

@stustison 
10, 1, etc were used as stubs, and are now refactored as constants.  These are currently not used, but may be implemented in the future as the stanag source is built.
Embed-dependency was removed.
Moved the mock server to sdk.

troymohl commented:
:+1:

shane-voisard commented:
Successful hero build of cca1b573ff179a8734f993c87e95bbeb6b6b5f79 rebased on 15e5cf59311a57c54d5796addd3cbb861eb1f809 .

bdeining commented:
After discussion, STANAG 4559 stuff is going to be moved to a different project.  This PR won't be merged, but will be left open for code review until the new project is created.

bdeining commented:
https://github.com/codice/alliance/pull/1
